File: LeetCode_Solutions/1171-remove-zero-sum-consecutive-nodes-from-linked-list/remove-zero-sum-consecutive-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        hm = {}
        pSum = 0
        hm[pSum] = dummy
        while head:
            pSum += head.val
            if pSum in hm:
                rem = hm[pSum].next
                SSum = pSum
                while rem != head:
                    SSum += rem.val
                    del hm[SSum]
                    rem = rem.next
                hm[pSum].next = head.next
            else:
                hm[pSum] = head
            head = head.next
        return dummy.next

File: LeetCode_Solutions/1267-remove-zero-sum-consecutive-nodes-from-linked-list/remove-zero-sum-consecutive-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode dummy = new ListNode();
        dummy.next = head;
        Map<Integer, ListNode> hm = new HashMap<>();
        int pSum = 0;
        hm.put(pSum, dummy);
        while (head != null) {
            pSum += head.val;
            if (hm.containsKey(pSum)) {
                ListNode rem = hm.get(pSum).next;
                int SSum = pSum;
                while (rem != head) {
                    SSum += rem.val;
                    hm.remove(SSum);
                    rem = rem.next;
                }
                hm.get(pSum).next = head.next;
            } else {
                hm.put(pSum, head);
            }
            head = head.next;
        }
        return dummy.next;
    }
}